libraries/riscv-csr/src/csr.rs

```diff
@@ -33,73 +69,272 @@ impl<T: IntLike, R: RegisterLongName> ReadWriteRiscvCsr<T, R> {
+            unsafe {
+                asm!("csrw {csr}, {rs}", rs = in(reg) val_to_set, csr = const PMPADDR4);
+            }
+        } else if self.value == PMPADDR5 {
+            unsafe {
+                asm!("csrw {csr}, {rs}", rs = in(reg) val_to_set, csr = const PMPADDR5);
+            }
+        } else {
+            panic!("Unsupported CSR write");
+        }
     }
 
     // Mock implementations for tests on Travis-CI.
     #[cfg(not(any(target_arch = "riscv32", target_os = "none")))]
-    pub fn get(&self) -> T {
+    pub fn get(&self) -> u32 {
         unimplemented!("reading RISC-V CSR {}", self.value)
     }
 
     #[cfg(not(any(target_arch = "riscv32", target_os = "none")))]
-    pub fn set(&self, _val_to_set: T) {
+    pub fn set(&self, _val_to_set: u32) {
         unimplemented!("writing RISC-V CSR {}", self.value)
     }
 
     #[inline]
-    pub fn read(&self, field: Field<T, R>) -> T {
+    pub fn read(&self, field: Field<u32, R>) -> u32 {
         (self.get() & (field.mask << field.shift)) >> field.shift
     }
 
     #[inline]
-    pub fn read_as_enum<E: TryFromValue<T, EnumType = E>>(&self, field: Field<T, R>) -> Option<E> {
-        let val: T = self.read(field);
+    pub fn read_as_enum<E: TryFromValue<u32, EnumType = E>>(
+        &self,
+        field: Field<u32, R>,
+    ) -> Option<E> {
+        let val: u32 = self.read(field);
 
         E::try_from(val)
     }
 
     #[inline]
-    pub fn extract(&self) -> LocalRegisterCopy<T, R> {
+    pub fn extract(&self) -> LocalRegisterCopy<u32, R> {
         LocalRegisterCopy::new(self.get())
     }
 
     #[inline]
-    pub fn write(&self, field: FieldValue<T, R>) {
+    pub fn write(&self, field: FieldValue<u32, R>) {
         self.set(field.value);
     }
 
     #[inline]
-    pub fn modify(&self, field: FieldValue<T, R>) {
-        let reg: T = self.get();
+    pub fn modify(&self, field: FieldValue<u32, R>) {
+        let reg: u32 = self.get();
         self.set((reg & !field.mask) | field.value);
     }
 
     #[inline]
-    pub fn modify_no_read(&self, original: LocalRegisterCopy<T, R>, field: FieldValue<T, R>) {
+    pub fn modify_no_read(&self, original: LocalRegisterCopy<u32, R>, field: FieldValue<u32, R>) {
         self.set((original.get() & !field.mask) | field.value);
     }
     #[inline]
-    pub fn is_set(&self, field: Field<T, R>) -> bool {
-        self.read(field) != T::zero()
+    pub fn is_set(&self, field: Field<u32, R>) -> bool {
+        self.read(field) != u32::zero()
     }
 
     #[inline]
-    pub fn matches_any(&self, field: FieldValue<T, R>) -> bool {
-        self.get() & field.mask != T::zero()
+    pub fn matches_any(&self, field: FieldValue<u32, R>) -> bool {
+        self.get() & field.mask != u32::zero()
     }
 
     #[inline]
-    pub fn matches_all(&self, field: FieldValue<T, R>) -> bool {
+    pub fn matches_all(&self, field: FieldValue<u32, R>) -> bool {
         self.get() & field.mask == field.value
     }
 }
```

Comment:
All of these functions look very similar to what's already available in `libraries/tock-register-interface/src/registers.rs`. Is there any reason not to use that?

Reply:
Reading a CSR requires an assembly instruction, not accessing a memory location.

Reply:
A lot of the logic still overlaps with methods such as `FieldValue::matches_all`. The only difference that I see is in the implementation of the `set` and `get` methods - all of the other methods are derived from it in the same manner as in the `libraries/tock-register-interface` library.

It looks like the `ReadAble` and `WriteAble` traits that I sketched in https://github.com/tock/tock/pull/1766#issue-405159759 could be beneficial here as well.